sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/Identifier.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalog.v2;
+
+import org.apache.spark.annotation.Experimental;
+
+/**
+ * Identifies an object in a catalog.
+ */
+@Experimental
+public interface Identifier {
```

Comment:
Shall we use a class directly? I don't see much value of using an interface here, as it has only one implementation.

Reply:
This allows us more flexibility than a single concrete class. Changing a class to an interface is not a binary compatible change, so using an interface is the right thing to do.

Reply:
Then I suggest we move the impl class to a private package like `org.apache.spark.sql.catalyst`. Also the static method should be moved to the impl class as well, as we only create it inside Spark.

Reply:
The implementation class is package-private. If we were to move it to a different package, we would need to make it public for the `of` factory method, which would increase its visibility, not decrease it.